odl/contrib/shearlab/shearlab_operator.py

```diff
@@ -0,0 +1,324 @@
+
+"""ODL integration with shearlab."""
+
+import odl
+import numpy as np
+from threading import Lock
+
+# Library for the shearlab library
+
+import julia
+import matplotlib.pyplot as plt
+from numpy import ceil
+from numpy.fft import fft2, ifft2, fftshift, ifftshift
+
+__all__ = ('ShearlabOperator',)
+
+
+class ShearlabOperator(odl.Operator):
+
+    """Shearlet transform using Shearlab.jl as backend.
+    This is the non-compact shearlet transform implemented using the Fourier
+    transform.
+    """
+
+    def __init__(self, space, num_scales):
+
+        """Initialize a new instance.
+        Parameters
+        ----------
+        space : `DiscreteLp`
+            The space on which the shearlet transform should act. Must be
+            two-dimensional.
+        num_scales : nonnegative `int`
+            The number of scales for the shearlet transform, higher numbers
+            mean better edge resolution but more computational burden.
+        Examples
+        --------
+        Create a 2d-shearlet transform:
+        >>> space = odl.uniform_discr([-1, -1], [1, 1], [128, 128])
+        >>> shearlet_transform = ShearlabOperator(space, num_scales=2)
+        """
+        self.shearlet_system = getshearletsystem2D(
+            space.shape[0], space.shape[1], num_scales)
+        range = space ** self.shearlet_system.nShearlets
+        self.mutex = Lock()
+        super(ShearlabOperator, self).__init__(space, range, True)
+
+    def _call(self, x):
+        """``self(x)``."""
+        with self.mutex:
+            result = sheardec2D(x, self.shearlet_system)
+            return np.moveaxis(result, -1, 0)
+
+    @property
+    def adjoint(self):
+        """The adjoint operator."""
+        op = self
+
+        class ShearlabOperatorAdjoint(odl.Operator):
+
+            """Adjoint of the shearlet transform.
+            Should not be used independently.
+            See Also
+            --------
+            odl.contrib.shearlab.ShearlabOperator
+            """
+
+            def __init__(self):
+                """Initialize a new instance.
+                Parameters
+                ----------
+                op : `ShearlabOperator`
+                    The operator which this should be the adjoint of.
+                """
+                super(ShearlabOperatorAdjoint, self).__init__(
+                    op.range, op.domain, True)
+
+            def _call(self, x):
+                """``self(x)``."""
+                with op.mutex:
+                    x = np.moveaxis(x, 0, -1)
+                    return sheardecadjoint2D(x, op.shearlet_system)
+
+            @property
+            def adjoint(self):
+                """The adjoint operator."""
+                return op
+
+            @property
+            def inverse(self):
+                """The inverse operator."""
+                op = self
+
+                class ShearlabOperatorAdjointInverse(odl.Operator):
+
+                    """\
```

Comment:
Please no backslashes in strings. The summary of this operator is mainly visible to developers, so you can keep it short, like "Inverse adjoint shearlet transform."